### PR TITLE
[native] Share RMSNorm forward and backward kernel launchers

### PR DESCRIPTION
**Pull request by slayton58**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #197011
* #197009
* __->__ #197008

The existing RMSNorm JIT adapter converts inputs to CuTe objects on each call
and finishes the weight gradient with ATen sum/cast operations. Add CuTeDSL
launchers for fp16, bf16, and fp32 at normalized widths 128, 512, 1024, 2048,
4096, and 8192 on SM90/SM100. Compile against fake tensor descriptors, pass
read-only tensors directly through TVM FFI, and finish partial weight gradients
with a dedicated CuTe reduction.

Start with `rmsnorm_kernels.py`, then the launch policy and adapter. The launchers
use the compilation target to select Quack configurations and compute weight
gradients only when requested. Other eligible inputs retain the existing
vendored launch path. Calls requesting neither gradient return without a launch.

Performance
-----------
Measured on NVIDIA B200 with CUDA 13.2, comparing this commit with its parent.
The 12-case benchmark explicitly selects the JIT or ATen route. Times are in
microseconds: eager latency is the warmed-up `Timer` median (at least 0.2 seconds
per case); GPU latency sums median kernel durations over 40 calls in a separate
profiler process. Selected results follow.

| Dtype | Shape | Operation | GPU before | GPU after | ATen GPU | Eager before | Eager after |
| --- | --- | --- | ---: | ---: | ---: | ---: | ---: |
| bfloat16 | 1 x 128 | backward | 14.74 | 3.17 | 3.47 | 94.93 | 68.10 |
| bfloat16 | 256 x 1024 | forward | 2.21 | 2.21 | 3.07 | 54.83 | 51.45 |
| bfloat16 | 256 x 1024 | backward | 14.21 | 8.06 | 5.25 | 96.49 | 69.43 |
| bfloat16 | 4096 x 1024 | backward | 16.75 | 19.17 | 19.63 | 95.84 | 70.03 |
| bfloat16 | 4096 x 8192 | backward | 52.37 | 64.27 | 96.62 | 93.90 | 70.28 |
| float32 | 8192 x 1024 | dw | 22.51 | 43.81 | 17.92 | 87.40 | 70.00 |

Backward computes both gradients; dw requests only the weight gradient.

Several larger backward cases regress in GPU time in this commit, despite
lower eager latency. #197011 addresses these regressions through grid and
reduction tuning and weight-only fallbacks.

Save the benchmark script below as `agent_space/publish-stacks/bench_commit.py`,
then run from `agent_space/`:

```bash
TORCH_DISABLE_NATIVE_AOT=1 QUACK_CACHE_DIR="$PWD/stack-assembly/jit-cache-stage2" python publish-stacks/bench_commit.py --tag shared --measure eager
TORCH_DISABLE_NATIVE_AOT=1 QUACK_CACHE_DIR="$PWD/stack-assembly/jit-cache-stage2" python publish-stacks/bench_commit.py --tag shared --measure gpu
```

<details>
<summary>Benchmark script: agent_space/publish-stacks/bench_commit.py</summary>

```python
import argparse
import contextlib
import hashlib
import json
import statistics
from collections import defaultdict
from pathlib import Path

import torch

parser = argparse.ArgumentParser()
parser.add_argument('--tag', required=True)
parser.add_argument('--measure', choices=('eager', 'gpu'), required=True)
args = parser.parse_args()
root = Path(__file__).resolve().parents[2]

@contextlib.contextmanager
def route(name):
    with torch.backends.python_native.cutedsl.disabled():
        if name == 'jit':
            torch.backends.python_native.cutedsl.enable()
        try:
            yield
        finally:
            if name == 'jit':
                torch.backends.python_native.cutedsl.disable()

cases = [
    (torch.bfloat16, 1, 128, 'forward'),
    (torch.bfloat16, 1, 128, 'backward'),
    (torch.bfloat16, 256, 1024, 'forward'),
    (torch.bfloat16, 256, 1024, 'backward'),
    (torch.bfloat16, 256, 1024, 'dw'),
    (torch.bfloat16, 4096, 1024, 'backward'),
    (torch.float32, 4096, 1024, 'backward'),
    (torch.bfloat16, 4096, 8192, 'forward'),
    (torch.bfloat16, 4096, 8192, 'backward'),
    (torch.bfloat16, 4096, 8192, 'dw'),
    (torch.bfloat16, 8192, 128, 'dw'),
    (torch.float32, 8192, 1024, 'dw'),
]
rows = []
for dtype, m, n, operation in cases:
    x = torch.randn(m, n, device='cuda', dtype=dtype)
    w = torch.randn(n, device='cuda', dtype=dtype)
    dy = torch.randn_like(x)
    with route('aten'):
        _, rstd = torch.ops.aten._fused_rms_norm(x, [n], w, 1e-5)
    def run():
        if operation == 'forward':
            return torch.ops.aten._fused_rms_norm(x, [n], w, 1e-5)
        return torch.ops.aten._fused_rms_norm_backward(
            dy, x, [n], rstd, w, [operation != 'dw', True]
        )
    for name in ('aten', 'jit'):
        with route(name):
            for _ in range(10):
                run()
            torch.cuda.synchronize()
            if args.measure == 'eager':
                from torch.utils.benchmark import Timer
                measurement = Timer(stmt='run()', globals={'run': run}).blocked_autorange(min_run_time=0.2)
                us = measurement.median * 1e6
            else:
                from torch.profiler import profile, ProfilerActivity
                with profile(activities=[ProfilerActivity.CUDA]) as prof:
                    for _ in range(40):
                        run()
                durations = defaultdict(list)
                for event in prof.events():
                    if event.device_type.name == 'CUDA':
                        durations[event.name].append(event.device_time_total)
                us = sum(statistics.median(values) * len(values) / 40 for values in durations.values())
        row = dict(dtype=str(dtype), m=m, n=n, operation=operation, route=name, us=us)
        rows.append(row)
        print(json.dumps(row), flush=True)
source = {
    str(path.relative_to(root)): hashlib.sha256(path.read_bytes()).hexdigest()
    for path in (root/'torch/_native/ops/norm').glob('*.py')
}
output = dict(tag=args.tag, measure=args.measure, gpu=torch.cuda.get_device_name(), cuda=torch.version.cuda, source_hashes=source, results=rows)
(root/f'agent_space/publish-stacks/{args.tag}-{args.measure}.json').write_text(json.dumps(output, indent=2)+'\n')
```
</details>

Test Plan
---------
On B200, 36 forward/backward cases and 77 existing routing/numerical tests
passed, including copy-on-write inputs. Commands run from `agent_space/`:

```bash
TORCH_DISABLE_NATIVE_AOT=1 QUACK_CACHE_DIR="$PWD/stack-assembly/jit-cache-stage2" python ../test/python_native/test_rmsnorm_cutedsl.py
TORCH_DISABLE_NATIVE_AOT=1 QUACK_CACHE_DIR="$PWD/stack-assembly/jit-cache-stage2" python ../test/test_nn.py TestFusedRMSNormOverrideRouting TestFusedRMSNormOverrideNumerics
```

Authored with an AI assistant.

### CODE REVIEW

**Comment by slayton58** [comment]
Only comment from local codex mock-review relates to invalid cache-reuse between commits on this stack -- given the stack should be merged as one, this is (imo) a non-issue. No other blockers or significant issues were raised.
